src/distribution_energy.cpp

```diff
@@ -0,0 +1,356 @@
+#include "openmc/distribution_energy.h"
+
+#include <algorithm> // for max, min, copy, move
+#include <cstddef>   // for size_t
+#include <iterator>  // for back_inserter
+
+#include "xtensor/xview.hpp"
+
+#include "openmc/endf.h"
+#include "openmc/hdf5_interface.h"
+#include "openmc/math_functions.h"
+#include "openmc/random_lcg.h"
+#include "openmc/search.h"
+
+namespace openmc {
+
+//==============================================================================
+// DiscretePhoton implementation
+//==============================================================================
+
+DiscretePhoton::DiscretePhoton(hid_t group)
+{
+  read_attribute(group, "primary_flag", primary_flag_);
+  read_attribute(group, "energy", energy_);
+  read_attribute(group, "atomic_weight_ratio", A_);
+}
+
+double DiscretePhoton::sample(double E) const
+{
+  if (primary_flag_ == 2) {
+    return energy_ + A_/(A_+ 1)*E;
+  } else {
+    return energy_;
+  }
+}
+
+//==============================================================================
+// LevelInelastic implementation
+//==============================================================================
+
+LevelInelastic::LevelInelastic(hid_t group)
+{
+  read_attribute(group, "threshold", threshold_);
```

Comment:
You could consider loading attribute names from a constant string attribute list to help improve documentation of all of the available attributes as well as preventing typos in the future.